### PR TITLE
Support token counting for OpenAI, DashScope, Gemini and HuggingFace models

### PR DESCRIPTION
**Pull request by DavdGao**
## Description

1. Add a new tokens module in AgentScope to handle token related events
2. Support to count tokens for OpenAI, DashScope, Gemini and HuggingFace models
3. Support to register token counting functions

## TODO
A more elegant way to register token counting functions for different models

## Checklist

Please check the following items before code is ready to be reviewed.

- [x]  Code has passed all tests
- [x]  Docstrings have been added/updated in Google Style
- [x]  Documentation has been updated
- [x]  Code is ready for review

### CODE REVIEW

File: src/agentscope/service/__init__.py [hunk 1]
```diff
@@ -66,7 +66,8 @@
 
 def get_help() -> None:
     """Get help message."""
-    help_msg = f"The following service are available:\n{__all__}"
+    tools = "\n - ".join(__all__)
```
Comment (zhijianma) [inline src/agentscope/service/__init__.py:69]: 1. 第一个元素会缺少 `-`
2. `ServiceResponse ` 等接口，是否要排出

File: src/agentscope/tokens.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    except ImportError as exc:
+        raise ImportError(
+            "The package `dashscope` is required for token counting "
+            "for Dashscope models.",
+        ) from exc
+
+    response = dashscope.Tokenization.call(
+        model=model_name,
+        messages=messages,
+        api_key=api_key or os.environ.get(DASHSCOPE_API_KEY_ENV),
```
Comment (zhijianma) [inline src/agentscope/tokens.py:279]: `DASHSCOPE_API_KEY_ENV`仅是该环境变量的一个存储名，实际为 `DASHSCOPE_API_KEY`，而且是dash对外公布的环境变量。建议去掉或者保持统一。

File: src/agentscope/tokens.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    Note this function will call the Dashscope API to count the tokens.
+    Refer to
+    https://help.aliyun.com/zh/dashscope/developer-reference/token-api?spm=5176.28197632.console-base_help.dexternal.1c407e06Y2bQVB&disableWebsiteRedirect=true
+    for more details.
+
+    Args:
+        model_name (`str`):
+            The name of the Dashscope model, e.g. "qwen-max".
+        messages (`list[dict[str, str]]`):
+            The list of messages, each message is a dict with the key 'text'.
```
Comment (zhijianma) [inline src/agentscope/tokens.py:259]: 该接口应该只支持dashscope的文本类的模型，是不是不会有`text` key？

File: src/agentscope/tokens.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+
+def count(model_name: str, messages: list[dict[str, str]]) -> int:
+    """Count the number of tokens for the given model and messages.
+
+    Args:
+        model_name (`str`):
+            The name of the model.
+        messages (`list[dict[str, str]]`):
+            The list of messages, each message is a dict with the key 'text'.
```
Comment (zhijianma) [inline src/agentscope/tokens.py:22]: `text` 这里应该是仅针对openai 的vl 模型吧？

File: src/agentscope/tokens.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+                "The type field currently only supports 'text' "
+                f"and 'image_url', but got {typ}.",
+            )
+    return num_tokens
+
+
+def count_openai_tokens(  # pylint: disable=too-many-branches
+    model_name: str,
+    messages: list[dict[str, str]],
+) -> int:
```
Comment (zhijianma) [inline src/agentscope/tokens.py:139]: 这里是否考虑增加一个license信息？

File: src/agentscope/tokens.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+            else:
+                raise TypeError(
+                    f"Invalid type {type(value)} in the {key} field.",
+                )
+
+            if key == "name":
+                num_tokens += tokens_per_name
+
+    num_tokens += 3  # every reply is primed with <|start|>assistant<|message|>
```
Comment (zhijianma) [inline src/agentscope/tokens.py:212]: 看上去 直接 给`num_tokens `初始化赋值为3也行

Comment (DavdGao) [reply src/agentscope/tokens.py]: Fixed.

File: src/agentscope/tokens.py (hunk 4 above)
Comment (DavdGao) [reply src/agentscope/tokens.py]: Fixed.

File: src/agentscope/tokens.py (hunk 3 above)
Comment (DavdGao) [reply src/agentscope/tokens.py:258]: Fixed.

File: src/agentscope/tokens.py (hunk 2 above)
Comment (DavdGao) [reply src/agentscope/tokens.py]: DONE

File: src/agentscope/tokens.py (hunk 5 above)
Comment (DavdGao) [reply src/agentscope/tokens.py:139]: DONE

**Review by zhijianma — APPROVED** [review]

**Pull request merged**